### PR TITLE
Mamba2: SSM State Caching for Efficient Incremental Generation 

### LINKED ISSUE #120: Mamba2: SSM State Caching for Efficient Incremental Generation
# Mamba2: SSM State Caching for Efficient Incremental Generation

## Motivation

Currently, `generate.py` recomputes the full sequence at every decoding step, giving O(n²) complexity. With state caching, we only need one forward pass per new token: O(n).

We can think of this as analogous to KV-caching in Transformers, but SSM state has a key advantage. KV-cache grows with sequence length. SSM state is fixed-size per layer: `(batch, heads, head_dim, state_size)`. The entire sequence history is compressed into this state, regardless of how long the sequence is.

## Current State

Already exists:
- `ssd_forward` accepts `initial_states` and `return_final_states`
- `Mamba2Model` threads per-layer states through

Missing:
- **Conv1d state caching**: Need to cache last `kernel_size - 1` values of `xBC` (post-projection, pre-conv). These get prepended before conv instead of zero-padding.
- **Cache object**: Unified structure holding SSM + conv states across layers
- **Generation loop**: Update to use cache instead of full recomputation

## Proposed Design

```python
@dataclasses.dataclass
class Mamba2Cache:
    ssm_states: list[jnp.ndarray]   # (batch, heads, head_dim, state_size) per layer
    conv_states: list[jnp.ndarray]  # (batch, d_conv_input, kernel_size - 1) per layer
                                    # where d_conv_input = intermediate_size + 2*state_size
```

Mixer signature becomes:
```python
def __call__(
    self, 
    hidden_states: jnp.ndarray, 
    cache: Mamba2Cache | None = None,
) -> tuple[jnp.ndarray, Mamba2Cache]:
```

Always returns `(output, cache)`. Pass `cache=None` on first call (prefill), then pass the returned cache on subsequent calls (decode).

## Tests

| Test | What it verifies |
|------|------------------|
| `test_cached_matches_full` | Process `[a,b,c]` full-sequence vs `[a,b]` then `[c]` with cache — logits for `c` must match |
| `test_conv_state_shapes` | Conv state is `(batch, intermediate_size + 2*state_size, kernel_size - 1)` |
| `test_ssm_state_shapes` | SSM state is `(batch, heads, head_dim, state_size)` per layer |

### PR DESCRIPTION
**Pull request by CosmoNaught**
Resolves #120 

**Reference**

Adds incremental-generation state caching for Mamba2 to avoid full-sequence recomputation at every decode step. The change introduces a unified cache that persists both SSM states (fixed-size per layer, analogous to Transformer KV-cache but constant memory w.r.t. sequence length) and the missing Conv1d rolling buffer (the last `kernel_size−1` projected inputs) so decoding can run in O(n) total compute instead of O(n²). The model/mixer API is updated to accept an optional cache on the first (prefill) call and then return an updated cache each step for fast token-by-token generation.

**Checklist**

- [x] I have read the **[Contribution Guidelines](https://github.com/jax-ml/bonsai/blob/main/CONTRIBUTING.md#contributing-a-model)** and used [pre-commit hooks](https://github.com/jax-ml/bonsai/blob/main/CONTRIBUTING.md#linting-and-type-checking) to format this commit.
- [x] I have added all the necessary **unit tests** for my change. (`run_model.py` for model usage, `test_outputs.py` and/or `model_validation_colab.ipynb` for quality).
- [x] **(If using an LLM)** I have carefully reviewed and removed all **superfluous comments** or unneeded, commented-out code. Only necessary and functional code remains.
- [x] I have signed the **[Contributor License Agreement (CLA)](https://cla.developers.google.com/about)**.

### CODE REVIEW

**Comment by chapman20j** [comment]
Hi @CosmoNaught. I left some small comments but looks good to me.

File: bonsai/models/mamba2/params.py [hunk 1]
```diff
@@ -86,6 +86,33 @@ def count_parameters(model: nnx.Module) -> int:
     return sum(p.size for p in jax.tree.leaves(params))
 
 
+def create_empty_cache(
```
[the remark this answers is not in the document]
Comment (CosmoNaught) [reply bonsai/models/mamba2/params.py]: Agreed, it makes sense to keep cache creation alongside the cache definition. All done see https://github.com/jax-ml/bonsai/pull/131/commits/441e1e64de1306a5258b8da5956e9db7c3d18c53
